### ISSUE TITLE
allShortestPaths gives *only* NULL distances for all resulting node pairs

### ISSUE THREAD

**Issue by mullinger** [neo4j-contrib/neo4j-graph-algorithms#256]
Problem was reported in Slack starting with the following comment.
https://neo4j-users.slack.com/archives/C5C4JRFK7/p1502125905314965

Graph Model:
`(System)-[c:Connected]-(System)`

c is conceptually bi-directional. I added a second reverse :Connected between all systems for tests
c has no property, for tests I added a dummy property 'distance' with value 1

The graph has around 5400 nodes in 3 small and one large disconnected subgraphs. The large graph contains around 5000 nodes which can all reach each other.

The result size is around 27 million distances.

My general query looks like this:
`call algo.allShortestPaths.stream('distance', {nodeQuery:'System', relationshipQuery:'Connected', defaultValue:1.0}) yield sourceNodeId, targetNodeId, distance
return sourceNodeId, targetNodeId, distance`

All variations of the query produce the correct number of results, but distances are always null
`call algo.allShortestPaths.stream('doesNotExist', {nodeQuery:'System', relationshipQuery:'Connected', defaultValue:1.0}) ...`
`call algo.allShortestPaths.stream('', {nodeQuery:'System', relationshipQuery:'Connected'}) ...`
`call algo.allShortestPaths.stream(null, {nodeQuery:'System', relationshipQuery:'Connected'}) ...`
...

> tomasi 10:30 PM
>i played around a bit
>10:30
>seems there is some bug
>10:30
>can you please report an issue

**Comment by mullinger** [comment]
Hrm, some more testing leads me to believe I have an error in my null filters:

This query does not filter nulls, and when ordered by distance returns a lot of nulls at the beginning of the query
`call algo.allShortestPaths.stream('distance', {nodeQuery:'System', relationshipQuery:'Connected'}) yield sourceNodeId, targetNodeId, distance
WIth sourceNodeId, targetNodeId, distance
where distance is not null
return sourceNodeId, targetNodeId, distance
limit 100000`
still gives rows 

`│853           │1297          │18        │

│853           │1298          │null      │

│853           │1299          │null      │`

While this gives no rows due to the wrong null-comparison
`call algo.allShortestPaths.stream('distance', {nodeQuery:'System', relationshipQuery:'Connected'}) yield sourceNodeId, targetNodeId, distance
WIth sourceNodeId, targetNodeId, distance
where distance <> null
return sourceNodeId, targetNodeId, distance
limit 100000`

with result
`(no changes, no records)`

**Comment by jexp** [comment]
@mullinger which version of the library do you use? I looked at the source of the latest version and it cannot generate null for the distance field.

**Comment by mullinger** [comment]
Hello @jexp ,
I am using Neo4j 3.2.2 with  graph-algorithms-algo-3.2.2.1.jar

**Comment by mullinger** [comment]
A simple example of two disconnected subgraphs gives the correct results
```Create (s1:Systemm {sid:1}), (s2:Systemm{sid:2}), (s3:Systemm{sid:3}), (s4:Systemm{sid:4}), (s5:Systemm{sid:5})```

```
Match (s1:Systemm {sid:1}), (s2:Systemm{sid:2}), (s3:Systemm{sid:3}), (s4:Systemm{sid:4}), (s5:Systemm{sid:5})
Create (s1)-[:Connectedd]->(s2), (s2)-[:Connectedd]->(s1), (s2)-[:Connectedd]->(s3), (s3)-[:Connectedd]->(s2), (s4)-[:Connectedd]->(s5), (s5)-[:Connectedd]->(s4)
```

`call algo.allShortestPaths.stream(null, {nodeQuery:'Systemm', relationshipQuery:'Connectedd'}) yield sourceNodeId, targetNodeId, distance WIth sourceNodeId, targetNodeId, distance where distance is not null return sourceNodeId, targetNodeId, distance`

No nulls, and correct distances

**Comment by mknblch** [comment]
FYI: we have 2 implementations for `algo.allShortestPaths.stream`. One is used **if a property-name is supplied** (`org.neo4j.graphalgo.impl.AllShortestPaths`), another one (`org.neo4j.graphalgo.impl.MSBFSAllShortestPaths`) is used **if property-name is null or empty**. Anyway.. both impls. should not be able to emit null's. Im going to test this

**Comment by mullinger** [comment]
@mknblch 
Yes, I checked the sourcecode and tried both variations. I did not see any differences

**Comment by jexp** [comment]
I think I found the issue. If you use cypher-shell which uses the java driver (not JS like the browser). 
It returns Infinity for the distance. It seems that the JS driver returns "null" for infinity.

Perhaps we should just filter those out and not return them, b/c there is no real path between those nodes.

```
| 0            | 2065         | Infinity           |
| 0            | 2066         | Infinity           |
| 0            | 2067         | Infinity           |
| 0            | 2068         | Infinity           |
| 0            | 2069         | Infinity           |
| 0            | 2070         | Infinity           |
| 0            | 2071         | Infinity           |
| 0            | 2072         | Infinity           |
| 0            | 2073         | Infinity           |
| 0            | 2074         | Infinity           |
```

**Comment by mknblch** [comment]
Good point! Imgoing to change this

**Comment by mullinger** [comment]
Oh, that is devious.

So filtering for null was succesful (because there were no nulls to begin with).
And filtering for "infinity" might have been succesful if I had known in the web-ui that null means infinity.

The "order by distance DESC" also showed a lot of nulls which would have been consistent with the sort order of null according to the documentation. But again, the nulls were "infinity"... 

Thank you for looking into this!

**Comment by mullinger** [comment]
Is there a way to filter "infinity" with cypher until the fixed algo-library is released?

**Comment by jexp** [comment]
Here is a build that you can use meanwhile:https://dl.dropboxusercontent.com/u/14493611/graph-algorithms-algo-3.2.2.2-SNAPSHOT.jar

**Comment by jexp** [comment]
@knutwalker can you add an algo.isInfinity function that users can use to filter out such results?

**Comment by knutwalker** [comment]
@jexp added those in #441
